internal/controller/hostpool_feedback_controller.go

```diff
@@ -0,0 +1,307 @@
+func (r *HostPoolFeedbackReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("hostpool-feedback").
+		For(&ckv1alpha1.HostPool{}, builder.WithPredicates(HostPoolNamespacePredicate(r.hostPoolNamespace))).
+		Complete(r)
+}
+
+// Reconcile is the implementation of the reconciler interface.
+func (r *HostPoolFeedbackReconciler) Reconcile(ctx context.Context, request ctrl.Request) (result ctrl.Result, err error) {
+	// Fetch the object to reconcile, and do nothing if it no longer exists:
+	object := &ckv1alpha1.HostPool{}
+	err = r.hubClient.Get(ctx, request.NamespacedName, object)
+	if err != nil {
+		err = clnt.IgnoreNotFound(err)
+		return //nolint:nakedret
+	}
+
+	// Get the identifier of the host pool from the labels. If this isn't present it means that the object wasn't
+	// created by the fulfillment service, so we ignore it.
+	hostPoolID, ok := object.Labels[cloudkitHostPoolIDLabel]
+	if !ok {
+		r.logger.Info(
+			"There is no label containing the host pool identifier, will ignore it",
+			"label", cloudkitHostPoolIDLabel,
+		)
+		return
+	}
+
+	// Check if the HostPool is being deleted before fetching from fulfillment service
+	if !object.ObjectMeta.DeletionTimestamp.IsZero() {
+		r.logger.Info(
+			"HostPool is being deleted, skipping feedback reconciliation",
+		)
+		return
+	}
+
+	// Fetch the host pool:
+	hostPool, err := r.fetchHostPool(ctx, hostPoolID)
+	if err != nil {
+		return
+	}
+
+	// Create a task to do the rest of the job, but using copies of the objects, so that we can later compare the
+	// before and after values and save only the objects that have changed.
+	t := &hostPoolFeedbackReconcilerTask{
+		r:        r,
+		object:   object,
+		hostPool: clone(hostPool),
+	}
+
+	result, err = t.handleUpdate(ctx)
+	if err != nil {
+		return
+	}
+	// Save the objects that have changed:
+	err = r.saveHostPool(ctx, hostPool, t.hostPool)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (r *HostPoolFeedbackReconciler) fetchHostPool(ctx context.Context, id string) (hostPool *privatev1.HostPool, err error) {
+	response, err := r.hostPoolsClient.Get(ctx, privatev1.HostPoolsGetRequest_builder{
+		Id: id,
+	}.Build())
+	if err != nil {
+		return
+	}
+	hostPool = response.GetObject()
+	if !hostPool.HasSpec() {
+		hostPool.SetSpec(&privatev1.HostPoolSpec{})
+	}
+	if !hostPool.HasStatus() {
+		hostPool.SetStatus(&privatev1.HostPoolStatus{})
+	}
+	return
+}
+
+func (r *HostPoolFeedbackReconciler) saveHostPool(ctx context.Context, before, after *privatev1.HostPool) error {
+	if !equal(after, before) {
+		_, err := r.hostPoolsClient.Update(ctx, privatev1.HostPoolsUpdateRequest_builder{
+			Object: after,
+		}.Build())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *hostPoolFeedbackReconcilerTask) handleUpdate(ctx context.Context) (result ctrl.Result, err error) {
+	err = t.syncConditions(ctx)
+	if err != nil {
+		return
+	}
```

Comment:
Does this code mean that if `syncConditions` has no errors, then the code below is never run?

Reply:
correct, note that `syncConditions` mutates conditions in the CR